File: Real-Time-Search-Workshop/javascript/code/lab2.js
/**
 * @fileoverview Basic JSON operations with the node-redis client
 * @maker Joey Whelan
 */

export class Lab2 {

    async run(client) {
        console.log('\n*** Lab 2 - Insert a simple KVP as a JSON object ***');
        let result = await client.json.set('ex1:1', '$', 'val');
        console.log(result);

        console.log('\n*** Lab 2 - Insert a single-property JSON object ***');
        result = await client.json.set('ex1:2', '$', {"field1": "val1"});
        console.log(result);

        console.log('\n*** Lab 2 - Insert a JSON object with multiple properties ***');
        result = await client.json.set('ex1:3', '$', {"field1": "val1", "field2": "val2"});
        console.log(result);

        console.log('\n*** Lab 2 - Insert a JSON object with multiple properties of different data types ***');
        result = await client.json.set('ex1:4', '$', {"field1": "val1", "field2": 2, "field3": true, "field4": null});
        console.log(result);

        console.log('\n*** Lab 2 - Insert a JSON object that contains an array ***');
        result = await client.json.set('ex1:5', '$', {"arr1": ["val1", "val2", "val3"]});
        console.log(result);

        console.log('\n*** Lab 2 - Insert a JSON object that contains a nested object ***');
        result = await client.json.set('ex1:6', '$', {"obj1": {"str1": "val1", "num2": 2}});
        console.log(result); 

        console.log('\n*** Lab 2 - Insert a JSON object with a mixture of property data types ***');
        result = await client.json.set('ex1:7', '$', {
            "str1": "val1", 
            "str2": "val2", 
            "arr1":[1,2,3,4], 
            "obj1": { "num1": 1, "arr2":["val1","val2", "val3"] }
        });
        console.log(result); 

        console.log('\n*** Lab 2 - Set and Fetch a simple JSON KVP ***');
        await client.json.set('ex2:1', '$', 'val');
        result = await client.json.get('ex2:1', '$');
        console.log(result); 

        console.log('\n*** Lab 2 - Set and Fetch a single property from a JSON object ***');
        await client.json.set('ex2:2', '$', {"field1": "val1"});
        result = await client.json.get('ex2:2', { path: '$.field1' });
        console.log(result); 

        console.log('\n*** Lab 2 - Fetch multiple properties ***');
        await client.json.set('ex2:3', '$', {"field1": "val1", "field2": "val2"});
        result = await client.json.get('ex2:3', { path: ['$.field1', '$.field2'] });
        console.log(result); 

        console.log('\n*** Lab 2 - Fetch a property nested in another JSON object ***');
        await client.json.set('ex2:4', '$', {"obj1": {"str1": "val1", "num2": 2}});
        result = await client.json.get('ex2:4', { path: '$.obj1.num2' });
        console.log(result); 

        console.log('\n*** Lab 2 - Fetch properties within an array and utilize array subscripting ***');
        await client.json.set('ex2:5', '$', {"str1": "val1", "str2": "val2", "arr1":[1,2,3,4], "obj1": {"num1": 1,"arr2":["val1","val2", "val3"]}});
        result = await client.json.get('ex2:5', { path: '$.obj1.arr2' });
        console.log(result); 
        result = await client.json.get('ex2:5', { path: '$.arr1[1]' });
        console.log(result);
        result = await client.json.get('ex2:5', { path: '$.obj1.arr2[0:2]' });
        console.log(result);
        result = await client.json.get('ex2:5', { path: '$.arr1[-2:]' });
        console.log(result);

        console.log('\n*** Lab 2 - Update an entire JSON object ***');
        await client.json.set('ex3:1', '$', {"field1": "val1"});
        await client.json.set('ex3:1', '$', {"foo": "bar"});
        result = await client.json.get('ex3:1');
        console.log(result); 

        console.log('\n*** Lab 2 - Update a single property within an object ***');
        await client.json.set('ex3:2', '$', {"field1": "val1", "field2": "val2"});
        await client.json.set('ex3:2', '$.field1', 'foo');
        result = await client.json.get('ex3:2');
        console.log(result); 

        console.log('\n*** Lab 2 - Update a property in an embedded JSON object ***');
        await client.json.set('ex3:3', '$', {"obj1": {"str1": "val1", "num2": 2}});
        await client.json.set('ex3:3', '$.obj1.num2', 3);
        result = await client.json.get('ex3:3');
        console.log(result); 

        console.log('\n*** Lab 2 - Update an item in an array via index ***');
        await client.json.set('ex3:4', '$', {"arr1": ["val1", "val2", "val3"]});
        await client.json.set('ex3:4', '$.arr1[0]', 'foo');
        result = await client.json.get('ex3:4');
        console.log(result); 

        console.log('\n*** Lab 2 - Delete entire object/key ***');
        await client.json.set('ex4:1', '$', {"field1": "val1"});
        await client.json.del('ex4:1');
        result = await client.json.get('ex4:1');
        console.log(result); 

        console.log('\n*** Lab 2 - Delete a single property from an object ***');
        await client.json.set('ex4:2', '$', {"field1": "val1", "field2": "val2"});
        await client.json.del('ex4:2', '$.field1');
        result = await client.json.get('ex4:2');
        console.log(result); 

        console.log('\n*** Lab 2 - Delete a property from an embedded object ***');
        await client.json.set('ex4:3', '$', {"obj1": {"str1": "val1", "num2": 2}});
        await client.json.del('ex4:3', '$.obj1.num2');
        result = await client.json.get('ex4:3');
        console.log(result); 

        console.log('\n*** Lab 2 - Delete a single item from an array ***');
        await client.json.set('ex4:4', '$', {"arr1": ["val1", "val2", "val3"]});
        await client.json.del('ex4:4', '$.arr1[0]');
        result = await client.json.get('ex4:4');
        console.log(result); 
    }
}

File: Real-Time-Query-Workshop/java/code/src/main/java/com/redis/queryworkshop/Lab2.java
/**
 * @fileoverview Basic JSON operations with the redis client
 * @maker Joey Whelan
*/

package com.redis.queryworkshop;
import org.json.JSONArray;
import org.json.JSONObject;
import java.util.Arrays;
import redis.clients.jedis.JedisPooled;
import redis.clients.jedis.json.Path;

public class Lab2 {
    public void run(JedisPooled client) {   
        System.out.println("\n*** Lab 2 - Insert a simple KVP as a JSON object ***");
        String res = client.jsonSet("ex1:1", new Path("$"),"val");
        System.out.println(res);

        System.out.println("\n*** Lab 2 - Insert a single-property JSON object ***");
        JSONObject jo = new JSONObject();
        jo.put("field1","val1");
        res = client.jsonSet("ex1:2", jo);
        System.out.println(res);

        System.out.println("\n*** Lab 2 - Insert a JSON object with multiple properties ***");
        jo = new JSONObject();
        jo.put("field1","val1");
        jo.put("field2", "val2");
        res = client.jsonSet("ex1:3", jo);
        System.out.println(res);
 
        System.out.println("\n*** Lab 2 - Insert a JSON object with multiple properties of different data types ***");
        jo = new JSONObject();
        jo.put("field1","val1");
        jo.put("field2", "val2");
        jo.put("field3", Boolean.TRUE);
        jo.put("field4", JSONObject.NULL);
        res = client.jsonSet("ex1:4", jo);
        System.out.println(res);

        System.out.println("\n*** Lab 2 - Insert a JSON object that contains an array ***");
        JSONArray ja = new JSONArray(Arrays.asList("val1", "val2", "val3"));
        res = client.jsonSet("ex1:5", ja);
        System.out.println(res);

        System.out.println("\n*** Lab 2 - Insert a JSON object that contains a nested object ***");
        JSONObject no = new JSONObject();
        no.put("str1","val1");
        no.put("num2", 2);
        jo = new JSONObject();
        jo.put("obj1", no);
        res = client.jsonSet("ex1:6", jo);
        System.out.println(res);

        System.out.println("\n*** Lab 2 - Insert a JSON object with a mixture of property data types ***");
        no = new JSONObject();
        no.put("num1", 1);
        no.put("arr2", new JSONArray(Arrays.asList("val1", "val2", "val3")));
        jo = new JSONObject();
        jo.put("str1", "val1");   
        jo.put("str2", "val2");
        jo.put("arr1", new JSONArray(Arrays.asList(1, 2, 3, 4)));
        jo.put("obj1", no);
        res = client.jsonSet("ex1:7", jo);
        System.out.println(res);
        
        System.out.println("\n*** Lab 2 - Set and Fetch a simple JSON KVP ***");
        client.jsonSet("ex2:1", new Path("$"),"val");
        res = client.jsonGetAsPlainString("ex2:1", new Path("$"));
        System.out.println(res); 

        System.out.println("\n*** Lab 2 - Set and Fetch a single property from a JSON object ***");
        jo = new JSONObject();
        jo.put("field1", "val1"); 
        client.jsonSet("ex2:2", jo);
        res = client.jsonGetAsPlainString("ex2:2", new Path("$"));
        System.out.println(res);

        System.out.println("\n*** Lab 2 - Fetch multiple properties ***");
        jo = new JSONObject();
        jo.put("field1", "val1"); 
        jo.put("field2", "val2");
        client.jsonSet("ex2:3", jo);
        Object ro = client.jsonGet("ex2:3", 
                        new Path("$.field1"), 
                        new Path("$.field2"));
        System.out.println(ro);         
 
        System.out.println("\n*** Lab 2 - Fetch a property nested in another JSON object ***");
        no = new JSONObject();
        no.put("str1", "val1");
        no.put("num2", 2);
        jo = new JSONObject();  
        jo.put("obj1", no); 
        client.jsonSet("ex2:4", jo);
        res = client.jsonGetAsPlainString("ex2:4", new Path("$.obj1.num2"));
        System.out.println(res);
            
        System.out.println("\n*** Lab 2 - Fetch properties within an array and utilize array subscripting ***");
        no = new JSONObject();
        no.put("num1", 1);
        no.put("arr2", new JSONArray(Arrays.asList("val1", "val2", "val3")));
        jo = new JSONObject();
        jo.put("str1", "val1");   
        jo.put("str2", "val2");
        jo.put("arr1", new JSONArray(Arrays.asList(1, 2, 3, 4)));
        jo.put("obj1", no);
        client.jsonSet("ex2:5", jo);
        res = client.jsonGetAsPlainString("ex2:5", new Path("$.obj1.arr2"));
        System.out.println(res);
        res = client.jsonGetAsPlainString("ex2:5", new Path("$.arr1[1]"));
        System.out.println(res);  
        res = client.jsonGetAsPlainString("ex2:5", new Path("$.obj1.arr2[0:2]"));
        System.out.println(res);     
        res = client.jsonGetAsPlainString("ex2:5", new Path("$.arr1[-2:]"));
        System.out.println(res); 
             
        System.out.println("\n*** Lab 2 - Update an entire JSON object ***");
        jo = new JSONObject();
        jo.put("field1", "val1");
        client.jsonSet("ex3:1", jo);
        jo = new JSONObject();
        jo.put("foo", "bar");
        client.jsonSet("ex3:1", jo);
        res = client.jsonGetAsPlainString("ex3:1", new Path("$"));
        System.out.println(res); 

        System.out.println("\n*** Lab 2 - Update a single property within an object ***");
        jo = new JSONObject();
        jo.put("field1", "val1");
        jo.put("field2", "val2"); 
        client.jsonSet("ex3:2", jo);   
        client.jsonSet("ex3:2", new Path("$.field1"), "foo");
        res = client.jsonGetAsPlainString("ex3:2", new Path("$"));
        System.out.println(res); 
 
        System.out.println("\n*** Lab 2 - Update a property in an embedded JSON object ***");
        no = new JSONObject();
        no.put("str1", "val1");
        no.put("num2", 2);
        jo = new JSONObject();
        jo.put("obj1", no);
        client.jsonSet("ex3:3", jo);    
        client.jsonSet("ex3:3", new Path("$.obj1.num2"), 3);    
        res = client.jsonGetAsPlainString("ex3:3", new Path("$"));
        System.out.println(res);  
            
        System.out.println("\n*** Lab 2 - Update an item in an array via index ***");
        jo = new JSONObject();
        jo.put("arr1", new JSONArray(Arrays.asList("val1", "val2", "val3")));    
        client.jsonSet("ex3:4", jo);    
        client.jsonSet("ex3:4", new Path("$.arr1[0]"), "foo");    
        res = client.jsonGetAsPlainString("ex3:4", new Path("$"));
        System.out.println(res); 
            
        System.out.println("\n*** Lab 2 - Delete entire object/key ***");
        jo = new JSONObject();
        jo.put("field1", "val1"); 
        client.jsonSet("ex4:1", jo); 
        client.jsonDel("ex4:1");
        res = client.jsonGetAsPlainString("ex4:1", new Path("$"));
        System.out.println(res); 

        System.out.println("\n*** Lab 2 - Delete a single property from an object ***");
        jo = new JSONObject();
        jo.put("field1", "val1");
        jo.put("field2", "val2");
        client.jsonSet("ex4:2", jo); 
        client.jsonDel("ex4:2", new Path("$.field1"));
        res = client.jsonGetAsPlainString("ex4:2", new Path("$"));
        System.out.println(res); 

        System.out.println("\n*** Lab 2 - Delete a property from an embedded object ***");
        no = new JSONObject();
        no.put("str1", "val1");
        no.put("num2", 2);
        jo = new JSONObject();
        jo.put("obj1", no); 
        client.jsonSet("ex4:3", jo); 
        client.jsonDel("ex4:3", new Path("$.obj1.num2"));
        res = client.jsonGetAsPlainString("ex4:3", new Path("$"));
        System.out.println(res);    

        System.out.println("\n*** Lab 2 - Delete a single item from an array ***");
        jo = new JSONObject();
        jo.put("arr1", new JSONArray(Arrays.asList("val1", "val2", "val3")));  
        client.jsonSet("ex4:4", jo); 
        client.jsonDel("ex4:4", new Path("$.arr1[0]"));
        res = client.jsonGetAsPlainString("ex4:4", new Path("$"));
        System.out.println(res); 
    }
}